src/markdown-to-inner-partitions.js

```diff
@@ -0,0 +1,214 @@
+import { PARTITION_TYPES } from './partition-types';
+
+export function generateInnerPartitions(markdownSubstring) {
+  if (markdownSubstring.length === 0) {
+    return null;
+  }
+  let partitions = [];
+  let currentIndex = 0;
+  let nextIndex = _findNextInnerStart(currentIndex, markdownSubstring);
+  if (nextIndex === markdownSubstring.length) {
+    return null;
+  }
+  let currentChar, close;
+  while (nextIndex < markdownSubstring.length) {
+    if (nextIndex !== currentIndex) {
+      partitions.push(_buildTextPartition(markdownSubstring.substring(currentIndex, nextIndex)));
+      currentIndex = nextIndex;
+    } else {
+      currentChar = markdownSubstring.charAt(currentIndex);
+      if (currentChar === '*') {
+        close = _findNextValidChar(currentChar, currentIndex + 1, markdownSubstring);
+        if (close > 0) {
+          partitions.push(_buildBoldPartition(markdownSubstring.substring(currentIndex + 1, close)));
+        }
+      } else if (currentChar === '_') {
+        close = _findNextValidChar(currentChar, currentIndex + 1, markdownSubstring);
+        if (close > 0) {
+          partitions.push(_buildItalicsPartition(markdownSubstring.substring(currentIndex + 1, close)));
+        }
+      } else if (currentChar === '[') {
+        close = markdownSubstring.indexOf(')', currentIndex);
+        partitions.push(_buildLinkPartition(markdownSubstring.substring(currentIndex, close + 1)));
+      } else if (currentChar === '{') {
+        close = markdownSubstring.indexOf(')', currentIndex);
+        partitions.push(_buildRelationOrColorPartition(markdownSubstring.substring(currentIndex, close + 1)));
+      }
+      currentIndex = close + 1;
+    }
+    nextIndex = _findNextInnerStart(currentIndex, markdownSubstring);
+  }
+  if (currentIndex !== nextIndex) {
+    partitions.push(_buildTextPartition(markdownSubstring.substring(currentIndex)));
+  }
+  return partitions;
+}
+
+function _buildBoldPartition(markdownSubstring) {
+  let partitions = generateInnerPartitions(markdownSubstring);
+  if (!partitions) {
+    return { type: PARTITION_TYPES.BOLD, value: markdownSubstring.replace(/\\/g, '') };
+  } else {
+    return { type: PARTITION_TYPES.BOLD, partitions };
+  }
+}
+
+function _buildItalicsPartition(markdownSubstring) {
+  let partitions = generateInnerPartitions(markdownSubstring);
+  if (!partitions) {
+    return { type: PARTITION_TYPES.ITALICS, value: markdownSubstring.replace(/\\/g, '') };
```

Comment:
You use `markdownSubstring.replace(/\\/g, '')` a lot.  Why not make it a helper function?  That way it'll make more sense anyway.  Like `removeEscapedDelimiter` or something.